File: app/src/main/java/me/imzack/app/end/view/fragment/SimpleGuidePageFragment.java
package me.imzack.app.end.view.fragment;

import android.os.Bundle;
import android.support.annotation.DrawableRes;
import android.support.annotation.StringRes;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import me.imzack.app.end.R;
import me.imzack.app.end.common.Constant;

import java.io.Serializable;

import butterknife.BindView;
import butterknife.ButterKnife;

public class SimpleGuidePageFragment extends BaseFragment {

    @BindView(R.id.image_subject)
    ImageView mSubjectImage;
    @BindView(R.id.text_title)
    TextView mTitleText;
    @BindView(R.id.text_dscpt)
    TextView mDscptText;
    @BindView(R.id.btn_action)
    Button mActionButton;

    private static final String ARG_IMAGE_RES_ID = "0";
    private static final String ARG_TITLE_RES_ID = "1";
    private static final String ARG_DSCPT_RES_ID = "2";
    private static final String ARG_BTN_TEXT_RES_ID = "3";
    private static final String ARG_BTN_CLK_LSNR = "4";

    private int mImageResId, mTitleResId, mDscptResId, mBtnTextResId;
    private OnButtonClickListener mBtnClkLsnr;

    public static SimpleGuidePageFragment newInstance(@DrawableRes int imageResId, @StringRes int titleResId,
                                                      @StringRes int descriptionResId, @StringRes int buttonTextResId,
                                                      OnButtonClickListener onButtonClickListener) {
        SimpleGuidePageFragment fragment = new SimpleGuidePageFragment();
        Bundle args = new Bundle();
        args.putInt(ARG_IMAGE_RES_ID, imageResId);
        args.putInt(ARG_TITLE_RES_ID, titleResId);
        args.putInt(ARG_DSCPT_RES_ID, descriptionResId);
        args.putInt(ARG_BTN_TEXT_RES_ID, buttonTextResId);
        args.putSerializable(ARG_BTN_CLK_LSNR, onButtonClickListener);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Bundle args = getArguments();
        if (args != null) {
            mImageResId = args.getInt(ARG_IMAGE_RES_ID);
            mTitleResId = args.getInt(ARG_TITLE_RES_ID);
            mDscptResId = args.getInt(ARG_DSCPT_RES_ID);
            mBtnTextResId = args.getInt(ARG_BTN_TEXT_RES_ID);
            mBtnClkLsnr = (OnButtonClickListener) args.getSerializable(ARG_BTN_CLK_LSNR);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_simple_guide_page, container, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ButterKnife.bind(this, view);

        if (mImageResId == Constant.UNDEFINED_RES_ID) {
            mSubjectImage.setVisibility(View.INVISIBLE);
        } else {
            mSubjectImage.setImageResource(mImageResId);
        }
        if (mTitleResId == Constant.UNDEFINED_RES_ID) {
            mTitleText.setVisibility(View.INVISIBLE);
        } else {
            mTitleText.setText(mTitleResId);
        }
        if (mDscptResId == Constant.UNDEFINED_RES_ID) {
            mDscptText.setVisibility(View.INVISIBLE);
        } else {
            mDscptText.setText(mDscptResId);
        }
        if (mBtnTextResId == Constant.UNDEFINED_RES_ID) {
            mActionButton.setVisibility(View.INVISIBLE);
        } else {
            mActionButton.setText(mBtnTextResId);
        }
        mActionButton.setOnClickListener(mBtnClkLsnr);
    }

    @Override
    public void onDetach() {
        super.onDetach();
    }

    public static class Builder {

        private int mImageResId, mTitleResId, mDscptResId, mBtnTextResId;
        private OnButtonClickListener mBtnClkLsnr;
        
        public Builder() {
            mImageResId = Constant.UNDEFINED_RES_ID;
            mTitleResId = Constant.UNDEFINED_RES_ID;
            mDscptResId = Constant.UNDEFINED_RES_ID;
            mBtnTextResId = Constant.UNDEFINED_RES_ID;
        }

        public Builder setImage(@DrawableRes int resId) {
            mImageResId = resId;
            return this;
        }

        public Builder setTitle(@StringRes int resId) {
            mTitleResId = resId;
            return this;
        }

        public Builder setDescription(@StringRes int resId) {
            mDscptResId = resId;
            return this;
        }

        public Builder setButton(@StringRes int resId, OnButtonClickListener listener) {
            mBtnTextResId = resId;
            mBtnClkLsnr = listener;
            return this;
        }

        public SimpleGuidePageFragment create() {
            return newInstance(mImageResId, mTitleResId, mDscptResId, mBtnTextResId, mBtnClkLsnr);
        }
    }

    public interface OnButtonClickListener extends View.OnClickListener, Serializable {

    }
}

File: app/src/main/kotlin/me/imzack/app/end/view/fragment/SimpleGuidePageFragment.kt
package me.imzack.app.end.view.fragment

import android.os.Bundle
import android.support.annotation.DrawableRes
import android.support.annotation.StringRes
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kotlinx.android.synthetic.main.fragment_simple_guide_page.*
import me.imzack.app.end.R
import java.io.Serializable

class SimpleGuidePageFragment : BaseFragment() {

    companion object {

        private val ARG_IMAGE_RES_ID = "0"
        private val ARG_TITLE_RES_ID = "1"
        private val ARG_DSCPT_RES_ID = "2"
        private val ARG_BTN_TEXT_RES_ID = "3"
        private val ARG_BTN_CLK_LSNR = "4"

        fun newInstance(@DrawableRes imageResId: Int, @StringRes titleResId: Int,
                        @StringRes descriptionResId: Int, @StringRes buttonTextResId: Int,
                        onButtonClickListener: OnButtonClickListener?): SimpleGuidePageFragment {
            val fragment = SimpleGuidePageFragment()
            val args = Bundle()
            args.putInt(ARG_IMAGE_RES_ID, imageResId)
            args.putInt(ARG_TITLE_RES_ID, titleResId)
            args.putInt(ARG_DSCPT_RES_ID, descriptionResId)
            args.putInt(ARG_BTN_TEXT_RES_ID, buttonTextResId)
            args.putSerializable(ARG_BTN_CLK_LSNR, onButtonClickListener)
            fragment.arguments = args
            return fragment
        }
    }

    private var mImageResId = 0
    private var mTitleResId = 0
    private var mDscptResId = 0
    private var mBtnTextResId = 0
    private var mBtnClkLsnr: OnButtonClickListener? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val args = arguments
        if (args != null) {
            mImageResId = args.getInt(ARG_IMAGE_RES_ID)
            mTitleResId = args.getInt(ARG_TITLE_RES_ID)
            mDscptResId = args.getInt(ARG_DSCPT_RES_ID)
            mBtnTextResId = args.getInt(ARG_BTN_TEXT_RES_ID)
            mBtnClkLsnr = args.getSerializable(ARG_BTN_CLK_LSNR) as OnButtonClickListener
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?) =
            inflater.inflate(R.layout.fragment_simple_guide_page, container, false)!!

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        if (mImageResId == 0) {
            image_subject.visibility = View.INVISIBLE
        } else {
            image_subject.setImageResource(mImageResId)
        }
        if (mTitleResId == 0) {
            text_title.visibility = View.INVISIBLE
        } else {
            text_title.setText(mTitleResId)
        }
        if (mDscptResId == 0) {
            text_dscpt.visibility = View.INVISIBLE
        } else {
            text_dscpt.setText(mDscptResId)
        }
        if (mBtnTextResId == 0) {
            btn_action.visibility = View.INVISIBLE
        } else {
            btn_action.setText(mBtnTextResId)
        }
        btn_action.setOnClickListener(mBtnClkLsnr)
    }

    override fun onDetach() {
        super.onDetach()
    }

    class Builder {

        private var mImageResId = 0
        private var mTitleResId = 0
        private var mDscptResId = 0
        private var mBtnTextResId = 0
        private var mBtnClkLsnr: OnButtonClickListener? = null

        fun setImage(@DrawableRes resId: Int): Builder {
            mImageResId = resId
            return this
        }

        fun setTitle(@StringRes resId: Int): Builder {
            mTitleResId = resId
            return this
        }

        fun setDescription(@StringRes resId: Int): Builder {
            mDscptResId = resId
            return this
        }

        fun setButton(@StringRes resId: Int, listener: OnButtonClickListener?): Builder {
            mBtnTextResId = resId
            mBtnClkLsnr = listener
            return this
        }

        fun create() = newInstance(mImageResId, mTitleResId, mDscptResId, mBtnTextResId, mBtnClkLsnr)
    }

    interface OnButtonClickListener : View.OnClickListener, Serializable
}
